README.md
Redi
----------

Pooled redis, add a layer of indirection between server pool and key ring

The idea comes from http://blog.zawodny.com/2011/02/26/redis-sharding-at-craigslist/

- - -
Install
----------
    gem install redi

Configure
----------
The configuration should look like a normal redis configuration with the addition of a buckets key.
This tells redi how many buckets it should hash keys to before mapping them to the configured server.

redi.yml:

    development:
      - :host: 192.168.0.10
        :port: 6379
        :db: 0
        :buckets: 0 - 64
      - :host: 192.168.0.11
        :port: 6380
        :db: 0
        :buckets: 65 - 127

In the example above, it is possible to scale the 2 configured servers up to 128 servers without 
re-keying.
- - -

Adding a new server can be done as follows:

* start a new server process, call it r3.
* identify buckets to move to r3 from existing server r2.
* setup r3 as slave to replicate from r2.
* update configuration to point buckets to r3

<pre>
production:
 - :host: 192.168.0.10 # r1
   :port: 6379
   :db: 0
   :buckets: 0 - 64
 - :host: 192.168.0.11 # r2
   :port: 6380
   :db: 0
   :buckets: 65 - 95
 - :host: 192.168.0.11 # r3
   :port: 6380
   :db: 0
   :buckets: 96 - 127
</pre>

* use bucket key prefixes to prune old keys from r2.

  How you do this can vary depending on your application, but something like the pseudo code below is the idea:

    96..127.times do|i| # NOTE: using redis here not redi as we want to talk to r2 explicitly
      redis.del(redis.keys("n#{i}*"))
    end


ChangeLog.md
0.0.9
-----
* ensure tests don't kill all dbs
* add explicit versions on redis and redis-namespace dependencies

0.0.8
-----
* Improve test runtime to not nuke all dbs @bleything

0.0.7
-----
* Complete redis interface @bleything
* Improved test coverage

0.0.6
------
* Reorganized project to be easier to navigate

0.0.5
------
* add del method
* better mock handling for tests


test/test_redi.rb
require 'test/unit'
require 'redi'
require 'yaml'

TEST_CONFIG = %(
  - :host: 127.0.0.1
    :port: 6379
    :db: 6
    :buckets: 0-64
  - :host: 127.0.0.1
    :port: 6379
    :db: 7
    :buckets: 65-127
)

class TestRedi < Test::Unit::TestCase

  def setup
    Redi.config = YAML.load( TEST_CONFIG )
    Redi.flushdb
  end

  def teardown
    Redi.flushdb
  end

  def test_unimplemented_commands
    assert_raises( NotImplementedError, /has not yet been implemented/ ) do
      Redi.brpoplpush 'source', 'destination', 'timeout'
    end
  end

  def test_redis_pool
    redis = Redi.pool.redis_by_key('me:foo:1')
    assert_equal "n25", redis.namespace
    redis.set("me:foo:1", "hello")
    assert_equal "hello", redis.get("me:foo:1")
  end

  def test_using_hash_ring_strategy
    node_class = Struct.new(:id, :to_s)
    buckets = []
    128.times do|i|
      buckets << node_class.new(i, "n#{i}")
    end

    servers = []
    2.times do|i|
      servers << node_class.new(i, "s#{i}") 
    end

    ring1 = Redis::HashRing.new(buckets)

    bucket2server = {}
    buckets.each do|bucket|
      bucket2server[bucket.to_s] = servers[bucket.id % servers.size].to_s
    end

    servers_used = {}

    buckets_used = {}
    6400.times do|i|
      key = "me:foo#{i}"
      bucket = ring1.get_node(key)
      index = bucket.to_s.gsub(/n/,'').to_i
      assert index < 128 && index >= 0, "bucket out of range: #{bucket}"
      server = bucket2server[bucket.to_s]
      servers_used[ server.to_s ] ||= 0
      servers_used[ server.to_s ] += 1
      buckets_used[ bucket.to_s ] ||= 0
      buckets_used[ bucket.to_s ] += 1
    end

    assert_equal 3228, servers_used["s0"], "when hashing 6400 times in sequence, s0 should have 3228 hits"
    assert_equal 3172, servers_used["s1"], "when hashing 6400 times in sequence, s1 should have 3172 hits"

  end
end


lib/redi.rb
require 'redi/pool'

class Redi

  ### these commands are complicated to distribute across a pool and so
  ### for now are unimplemented. Translation: we are lazy.
  UNIMPLEMENTED_COMMANDS = %w[
    keys move object randomkey rename renamenx eval
    mget mset msetnx
    brpoplpush rpoplpush
    sdiff sdiffstore sinter sinterstore smove sunion sunionstore
    zinterstore zunionstore
    psubscribe publish punsubscribe subscribe unsubscribe
    discard exec multi unwatch watch
    auth echo ping quit select
    bgrewriteaof bgsave config dbsize debug info lastsave monitor save shutdown slaveof slowlog sync
  ]

  ### raise exceptions on unimplemented/unknown commands, delegate
  ### everything else down to the actual Redis connections
  def self.method_missing( cmd, *args )
    if UNIMPLEMENTED_COMMANDS.include?( cmd.to_s )
      raise NotImplementedError, "#{cmd} has not yet been implemented. Patches welcome!"
    end

    pool.redis_by_key( args.first ).send( cmd, *args )
  end

  def self.flushdb
    pool.flushdb
  end

  def self.flushall
    pool.flushall
  end

  def self.pool
    @pool ||= Pool.new(self.config)
  end

  def self.config=(config)
    @config = config
  end

  def self.config
    @config
  end

end


lib/redi/pool.rb
require 'rubygems'

require 'redis'
require 'redis/hash_ring'
require 'redis/namespace'

# provide a key to name to host:port mapping
#
#  should have a larger keyspace than servers, this allows scaling up the servers without changing the keyspace mapping
#
# sample configuration:
#
# - :host:
#   :port:
#   :db:
#   :buckets: 0 - 64
# - :host:
#   :port:
#   :db:
#   :buckets: 65 - 127
#
class Redi
  class Pool
    attr_reader :keyspace, :servers
    def initialize(config)
      key_type = Struct.new(:id, :to_s)

      # build server pool
      @bucket2server = {}
      buckets = []
      @servers = config.map {|cfg|
        bucket_range = cfg.delete(:buckets)
        s, e = bucket_range.split('-').map {|n| n.to_i }
        conn = Redis.new(cfg)
        (s..e).each do|i|
          bucket_name = "n#{i}"
          buckets << key_type.new(i, bucket_name)
          @bucket2server[bucket_name] = conn
        end
        conn
      }

      # create the keyring to map redis keys to buckets
      @keyring  = Redis::HashRing.new(buckets)
    end

    def qualified_key_for(key)
      bucket = @keyring.get_node(key)
      "#{bucket.to_s}:#{key}"
    end

    def redis_by_key(key)
      bucket = @keyring.get_node(key)
      Redis::Namespace.new(bucket.to_s, :redis => @bucket2server[bucket.to_s])
    end

    def flushall
      @servers.map {|s| s.flushall }
    end

    def flushdb
      @servers.each {|s| s.flushdb }
    end

  end
end


lib/redis/ring.rb
require 'rubygems' if __FILE__ == $0
require 'zlib'
require 'yaml'
require 'redis'
require 'redis/hash_ring'
require 'redis/namespace'

class Redis
class Ring

  def self.get(key)
    pool.redis_by_key(key).get(key)
  end

  def self.set(key, val)
    pool.redis_by_key(key).set(key, val)
  end

  def self.pool
    @pool ||= Pool.new(self.config) #Redis.new(self.config)
  end

  def self.config
    @config ||= YAML.load_file(File.join(RAILS_ROOT,'config/me.yml'))[RAILS_ENV]
  end

  # provide a key to name to host:port mapping
  #
  #  should have a larger keyspace than servers, this allows scaling up the servers without changing the keyspace mapping
  #
  # sample configuration:
  #
  # :keyspace:
  #   - :n0: 0
  #   - :n1: 1
  #   - :n2: 0
  #   - :n3: 1
  #   - :n4: 0
  # :servers:
  #   - :host:
  #     :port:
  #     :db:
  #   - :host:
  #     :port:
  #     :db:
  #
  class Pool
    attr_reader :keyspace, :servers
    def initialize(config)
      @key_type = Struct.new(:id, :to_s)
      keyspace = config[:keyspace]

      # create bucket set
      @buckets = keyspace.each_with_index.map {|k,i| @key_type.new(i, k.keys.first) }

      # build server pool
      @servers = config[:servers].map {|cfg| Redis.new(cfg) }

      # create the mapping from bucket to server
      @bucket2server = {}
      keyspace.each do|ks|
        @bucket2server[ks.keys.first] = @servers[ks.values.first]
      end

      # create the keyring to map redis keys to buckets
      @keyring  = Redis::HashRing.new(@buckets)
    end

    def redis_by_key(key)
      bucket = @keyring.get_node(key)
      Redis::Namespace.new(bucket.to_s, :redis => @bucket2server[bucket.to_s])
    end

  end

  class Mock
    def initialize
      @store = {}
    end

    def get(key)
      @store[key]
    end

    def set(key, val)
      @store[key] = val
    end

    def mget(*keys)
      keys.map {|k| get(k) }
    end

    def flushall
      @store = {}
    end

  end

end
end

if __FILE__ == $0
require 'rubygems'
require 'test/unit'

TEST_CONFIG = %(
  :keyspace:
    - :n0: 0
    - :n1: 1
    - :n2: 0
    - :n3: 1
    - :n4: 0
    - :n5: 1
    - :n6: 0
    - :n7: 1
    - :n8: 0
    - :n9: 1
    - :n10: 0
    - :n11: 1
    - :n12: 0
    - :n13: 1
    - :n14: 0
    - :n15: 1
    - :n16: 0
    - :n17: 1
    - :n18: 0
    - :n19: 1
    - :n20: 0
    - :n21: 1
    - :n22: 0
    - :n23: 1
    - :n24: 0
    - :n25: 1
    - :n26: 0
    - :n27: 1
    - :n28: 0
    - :n29: 1
    - :n30: 0
    - :n31: 1
    - :n32: 0
  :servers:
    - :host: 127.0.0.1
      :port: 6379
      :db: 0
    - :host: 127.0.0.1
      :port: 6379
      :db: 1
)
  class TestIt < Test::Unit::TestCase

    def test_redis_pool
      pool = Redis::Ring::Pool.new(YAML.load(TEST_CONFIG))
      redis = pool.redis_by_key('me:foo:1')
      assert_equal :n25, redis.namespace
      redis.flushall
      redis.set("me:foo:1", "hello")
      assert_equal "hello", redis.get("me:foo:1")
    end

    def test_using_hash_ring_strategy
      return
      node_class = Struct.new(:id, :to_s)
      buckets = []
      128.times do|i|
        buckets << node_class.new(i, "n#{i}")
      end

      servers = []
      2.times do|i|
        servers << node_class.new(i, "s#{i}") 
      end

      ring1 = Redis::HashRing.new(buckets)

      bucket2server = {}
      buckets.each do|bucket|
        bucket2server[bucket.to_s] = servers[bucket.id % servers.size].to_s
      end

      puts bucket2server.inspect

      servers_used = {}

      buckets_used = {}
      6400.times do|i|
        key = "me:foo#{i}"
        bucket = ring1.get_node(key)
        server = bucket2server[bucket.to_s]#ring2.get_node(name.to_s)
        #puts "#{key} -> #{name} -> #{server}"
        servers_used[ server.to_s ] ||= 0
        servers_used[ server.to_s ] += 1
        buckets_used[ bucket.to_s ] ||= 0
        buckets_used[ bucket.to_s ] += 1
      end
      puts servers_used.inspect
      puts buckets_used.inspect
    end
  end
end
